### PR TITLE
Enable Flash Attention in recompute and causal modes (#21)

### PR DESCRIPTION
**Pull request by wszczurekhabana**
Cherry-pick from: https://github.com/HabanaAI/optimum-habana-fork/pull/21

Original description:
This is a follow-up on: https://github.com/huggingface/optimum-habana/pull/623

where main issue is that when running with Flash Attention in causal mode (required for performance and memory optimizations on 1st token) it will generate a triangular attention mask - same on each batch of the input data.
If we have more than one batch of sequences that have different lengths, tokenizer will include a padding for those sentences that have smaller sequence lengths than 'max input tokens'. In the case of Flash Attention in causal mode, where triangular attention mask is applied to all the inputs, this will effectively mean that padding tokens are also attended to, which will in turn result in junk output being generated

This PR propagates the modes of running to run_generation.py so that optimization can still be used in case of the same length inputs in a batch. This is controlled through: --flash_attention_recompute and --flash_attention_causal_mask.

Additionally this PR provides a way to pass real input data to the model from Project Gutenberg Books for easier testing of large sequence lengths.

Below are the throughput measurements for different ratios of prompt to max seq length:

| Ratio       | Max input tokens | Max new tokens  | Batch size  | Throughput [tokens/s] |
| ---------- | ------------------- | ------------------ | ------------- | ------------- |
| 97%         | 31744                  | 1042                    | 12 | 85.54 |
| 75%         | 24576                  |  8192                   | 16 | 336.29 |
| 50%         | 16384                  | 16384                  | 24 | 521.39 |
| 25%         | 8192                    |  24576                 | 36 | 708.42 |

example of the command tested:
`
python ../gaudi_spawn.py --use_deepspeed --world_size 8 run_generation.py \
    --num_beams 1  --attn_softmax_bf16 --model_name_or_path meta-llama/Llama-2-70b-hf \
    --warmup 2 --n_iterations 3 --use_hpu_graphs --use_kv_cache --max_input_tokens 31744 --max_new_tokens 1042 --bf16 --batch_size 12 --reuse_cache --trim_logits --limit_hpu_graphs --use_flash_attention --flash_attention_recompute --flash_attention_causal_mask --book_source`
    
    
Results on finetuning:

No Flash Attention:
'train_runtime': 2499.5658, 'train_samples_per_second': 2.626

Flash Attention:
'train_runtime': 2487.0323, 'train_samples_per_second': 2.636

Flash Attention Causal:
'train_runtime': 2449.3563, 'train_samples_per_second': 2.686

### CODE REVIEW

File: optimum/habana/transformers/generation/utils.py [hunk 1]
```diff
@@ -726,6 +726,9 @@ def generate(
         # determine whether flash attention needs to be used
         model_kwargs["use_flash_attention"] = generation_config.use_flash_attention
         model_kwargs["flash_attention_recompute"] = True if generation_config.flash_attention_recompute else False
+        model_kwargs["flash_attention_causal_mask"] = True if generation_config.flash_attention_causal_mask else False
+        model_kwargs["use_fused_rope"] = False if not generation_config.use_fused_rope else True
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:730]: Let's remove line 730 as `use_fused_rope` is mostly a workaround for the LoRA example at the moment

File: examples/text-generation/run_generation.py [hunk 2]
```diff
@@ -272,6 +287,45 @@ def main():
...
+                        prompt += word + " "
+                        if counter == prompt_size:
+                            return [prompt] * args.batch_size
+
+            book_ids = [
+                2701,  # Moby Dick; Or, The Whale
+                1513,  # Romeo and Juliet
+                1342,  # Pride and Prejudice
+            ]
+            input_sentences = assemble_prompt(prompt_size=args.max_input_tokens, book_path=download_book(book_ids[0]))
```
Comment (regisss) [inline examples/text-generation/run_generation.py:328]: Isn't it the same as https://huggingface.co/datasets/manu/project_gutenberg?

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -406,10 +408,15 @@ def pre_attn_forward(
                     )
             else:
                 # first token
-                with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
-                    attn_output = FusedSDPA.apply(
-                        query_states, key_states, value_states, attention_mask, 0.0, False, None
-                    )
+                if flash_attention_causal_mask:
+                    # causal masking on first token requires inputs to be of the same lenght
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:412]:
```suggestion
                    # causal masking on first token requires inputs to be of the same length
```

File: examples/text-generation/README.md [hunk 4]
```diff
@@ -354,6 +354,30 @@ QUANT_CONFIG=./quantization_config/maxabs_quant.json python ../gaudi_spawn.py \
 ```
 `--fp8` is required to enable quantization in fp8.
 
+### Using Habana Flash Attention
+
+Habana Flash Attention addresses large sequence lenghts on prompt stage of inference. Using causal attention mask on prompt stage requires input sequences in batch to be of the same length, but can provide a memory saving, thus enabling higher batch sizes.
```
Comment (regisss) [inline examples/text-generation/README.md:359]:
```suggestion
Habana Flash Attention addresses large sequence lengths on prompt stage of inference. Using causal attention mask on prompt stage requires input sequences in batch to be of the same length, but can provide a memory saving, thus enabling higher batch sizes.
```

File: examples/text-generation/README.md [hunk 5]
```diff
@@ -354,6 +354,30 @@ QUANT_CONFIG=./quantization_config/maxabs_quant.json python ../gaudi_spawn.py \
...
+
+Below example uses `flash_attention_recompute` mode in order to reduce memory consumption on prompt stage. Additionally since all sequences in a batch are of the same lenght it uses `flash_attention_causal_mask` which will further improve performance by taking advantage of specific lower-diagonal shape of inputs to softmax operation.
```
Comment (regisss) [inline examples/text-generation/README.md:361]:
```suggestion
Below example uses `flash_attention_recompute` mode in order to reduce memory consumption on prompt stage. Additionally since all sequences in a batch are of the same length it uses `flash_attention_causal_mask` which will further improve performance by taking advantage of specific lower-diagonal shape of inputs to softmax operation.
```

**Review by regisss — COMMENTED** [review]
Why adding `use_fused_rope` in the changes of this PR? You need it for recompute and causal mode?

File: optimum/habana/transformers/generation/utils.py (hunk 1 above)
Comment (wszczurekhabana) [reply optimum/habana/transformers/generation/utils.py]: Not needed, I removed it in second commit.

File: examples/text-generation/run_generation.py (hunk 2 above)
Comment (wszczurekhabana) [reply examples/text-generation/run_generation.py:328]: It's the same source, but downloading the whole dataset takes a lot of time and it's not necessary for this test, so I've decided to download individual book only.

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 3 above)
Comment (wszczurekhabana) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Fixed in second commit.

File: examples/text-generation/README.md (hunk 4 above)
Comment (wszczurekhabana) [reply examples/text-generation/README.md]: Fixed in second commit.

File: optimum/habana/transformers/trainer.py [hunk 6]
```diff
@@ -925,6 +925,10 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args):
                             inputs["use_flash_attention"] = True
                         if self.model.generation_config.flash_attention_recompute:
                             inputs["flash_attention_recompute"] = True
+                        if self.model.generation_config.flash_attention_causal_mask:
+                            inputs["flash_attention_causal_mask"] = True
+                        if not self.model.generation_config.use_fused_rope:
+                            inputs["use_fused_rope"] = False
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:931]: Let's remove these two lines too and then it will be good to merge!

Comment (wszczurekhabana) [reply optimum/habana/transformers/trainer.py]: Removed in 3rd commit.

**Comment by regisss** [comment]
@wszczurekhabana I changed the target of this PR to main because we first need to merge this change there and then I'll cherry-pick it in the release branch.
Can you make sure your main branch is up to date, run `git merge main` in your working branch and then solve the merge conflicts please?

File: optimum/habana/version.py [hunk 7]
```diff
@@ -13,4 +13,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-__version__ = "1.11.0.dev0"
+__version__ = "1.11.0"
```
Comment (regisss) [inline optimum/habana/version.py:16]: Please remove this change

File: setup.py [hunk 8]
```diff
@@ -49,7 +49,7 @@
 
 QUALITY_REQUIRES = [
     "ruff",
-    "hf_doc_builder @ git+https://github.com/huggingface/doc-builder.git",
+    "hf_doc_builder",
```
Comment (regisss) [inline setup.py:52]: Same

**Review by regisss — APPROVED** [review]
